### PR TITLE
Pv2 175 process provider payment for small employers

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.ProviderPayments.Application/Infrastructure/ioc/AutoMapperModule.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using SFA.DAS.Payments.Application.Infrastructure.Ioc;
+using Module = Autofac.Module;
+
+namespace SFA.DAS.Payments.ProviderPayments.Application.Infrastructure.ioc
+{
+    public class AutoMapperModule : Module
+    {
+        protected override void Load(ContainerBuilder builder)
+        {
+            //TODO: Refactor to make it it more reusable
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Infrastructure/ioc/AutoMapperModule.cs]: is this going to happen?

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/IIlrSubmissionService.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using SFA.DAS.Payments.EarningEvents.Messages.Events;
+
+namespace SFA.DAS.Payments.ProviderPayments.Application.Services
+{
+    public interface IIlrSubmissionService
+    {
+        Task HandleIlrSubMission(IlrSubmittedEvent message, CancellationToken cancellationToken);
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/IIlrSubmissionService.cs]: Submission, not SubMission

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/IProviderPaymentsService.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿using SFA.DAS.Payments.Model.Core.Entities;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.Payments.ProviderPayments.Application.Services
+{
+    public interface IProviderPaymentsService
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/IProviderPaymentsService.cs]: There is another IProviderPaymentsService interface in SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService.Interfaces that implements the same as this interface. Can they be refactored?

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/IlrSubmissionService.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            IValidateIlrSubmission validateIlrSubmission,
+            IPaymentLogger paymentLogger)
+        {
+            this.providerPaymentsRepository = providerPaymentsRepository;
+            this.ilrSubmittedEventCache = ilrSubmittedEventCache;
+            this.validateIlrSubmission = validateIlrSubmission;
+            this.paymentLogger = paymentLogger;
+        }
+
+        public async Task HandleIlrSubMission(IlrSubmittedEvent message, CancellationToken cancellationToken)
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/IlrSubmissionService.cs]: rename as per interface to HandleIlrSubmission

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            this.ilrSubmittedEventCache = ilrSubmittedEventCache;
+        }
+
+        public async Task<List<PaymentModel>> GetMonthEndPayments(short collectionYear, byte collectionPeriod, long ukprn, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var payments = await providerPaymentsRepository.GetMonthEndPayments(collectionYear, collectionPeriod, ukprn, cancellationToken);
+            return payments.ToList();
+        }
+
+        private async Task<IlrSubmittedEvent> GetCurrentIlrSubmissionEvent(long ukprn, CancellationToken cancellationToken)
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.ProviderPayments.Application/Services/MonthEndService.cs]: this doesn't appear to be used

File: src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService.Interfaces/IProviderPaymentsService.cs [hunk 6]
```diff
@@ -9,8 +12,10 @@ namespace SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService.Interfaces
 {
     public interface IProviderPaymentsService : IActor
     {
-        Task<int> GetCountAsync(CancellationToken cancellationToken);
+        Task ProcessPayment(PaymentModel message, CancellationToken cancellationToken);
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.ProviderPayments.ProviderPaymentsService.Interfaces/IProviderPaymentsService.cs]: There is another interface with the same name in SFA.DAS.Payments.ProviderPayments.Application.Services that has the ProcessPayment signature in. Can it be refactored?

File: src/SFA.DAS.Payments.AcceptanceTests.Core/Automation/TestSession.cs [hunk 7]
```diff
@@ -13,7 +13,7 @@ public class TestSession
         public long Ukprn { get; }
         public List<Learner> Learners { get; }
         public Learner Learner => Learners.FirstOrDefault();
-        public string JobId { get; }
+        public long JobId { get; }
         //private static ConcurrentDictionary<string, ConcurrentBag<TestSession>> Sessions { get;  } = new ConcurrentDictionary<string, ConcurrentBag<TestSession>>();  //TODO: will need to be refactored at some point
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.Core/Automation/TestSession.cs]: This has been changed in End-to-End test branch will leave it for now.

File: src/SFA.DAS.Payments.AcceptanceTests.Core/Automation/TestSession.cs [hunk 8]
```diff
@@ -24,19 +24,19 @@ public TestSession()
...
                 .RuleFor(course => course.AgreedPrice, 15000);
-            
+
             SessionId = Guid.NewGuid().ToString();
             random = new Random(Guid.NewGuid().GetHashCode());
             Ukprn = GenerateId("ukprn");
-            Learners = new List<Learner>();
-            JobId = Guid.NewGuid().ToString("N");
+            Learners = new List<Learner> { GenerateLearner() };
+            JobId = GenerateId("JobId");
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.Core/Automation/TestSession.cs]: That is right, LearnAimRef  is a property of Course class which is a part of Acceptance Test Data

File: src/SFA.DAS.Payments.FundingSource.Messages/Events/FundingSourcePaymentEvent.cs [hunk 9]
```diff
@@ -1,29 +1,26 @@
-﻿using System;
+﻿using SFA.DAS.Payments.Model.Core;
+using SFA.DAS.Payments.Model.Core.OnProgramme;
+using System;
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.FundingSource.Messages/Events/FundingSourcePaymentEvent.cs]: The namespace should follow the directory hierarchy from the top - down

File: src/SFA.DAS.Payments.Messages.Core/Events/IPaymentsEvent.cs [hunk 10]
```diff
@@ -9,7 +9,7 @@ public interface IPaymentsEvent : IPaymentsMessage
         long Ukprn { get; }
         Learner Learner { get; }
         LearningAim LearningAim { get; }
-     
-   
+        DateTime IlrSubmissionDateTime { get; }
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.Messages.Core/Events/IPaymentsEvent.cs]: They are used by PaymentsEvent class that implements IPaymentsEvent

File: src/SFA.DAS.Payments.Messages.Core/IPaymentsMessage.cs [hunk 11]
```diff
@@ -2,6 +2,6 @@
 {
     public interface IPaymentsMessage
     {
-        string JobId { get; }
+        long JobId { get; }
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.Messages.Core/IPaymentsMessage.cs]: It is used and has 61 references

File: src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/Steps/ProviderPaymentsStepsBase.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            {
+                case FundingSourceType.CoInvestedSfa:
+                    paymentEvent = new SfaCoInvestedFundingSourcePaymentEvent();
+                    break;
+                case FundingSourceType.CoInvestedEmployer:
+                    paymentEvent = new EmployerCoInvestedFundingSourcePaymentEvent();
+                    break;
+                default:
+                    //TODO Implement other FundingSourceTypes
+                    throw new NotImplementedException("Unhandled Funding Source Type");
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.AcceptanceTests/Steps/ProviderPaymentsStepsBase.cs]: Yes as we want it to fail early

File: src/SFA.DAS.Payments.ProviderPayments.Application/Infrastructure/ioc/AutoMapperModule.cs (hunk 1 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Infrastructure/ioc/AutoMapperModule.cs]: I have remove the comment

File: src/SFA.DAS.Payments.ProviderPayments.Application/Services/IProviderPaymentsService.cs (hunk 3 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.ProviderPayments.Application/Services/IProviderPaymentsService.cs]: They serve a completely different purpose

**Pull request merged**